### PR TITLE
tests: fix some unstable test

### PR DESCRIPTION
**Pull request by Ehco1996**
### What problem does this PR solve? 

fix some unstable test [jenkins](https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/230/pipeline/980)

* ha
* gtid
* sharding

### What is changed and how it works?
- sleep in it test cace 

### Check List 

Tests 
 - Integration test
 - No code

Code changes
 - None

Side effects
- None

Related changes
 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by lance6716** [comment]
sleep (for a long interval) is not a good way, you could check + sleep for a short time + retry

**Comment by Ehco1996** [comment]
> sleep (for a long interval) is not a good way, you could check + sleep for a short time + retry

i can remove sleep in the `gitid` test case

but in `ha` case still need to sleep, because there is two master node in `initial-cluster` cfg , and i need mater1  to became the leader.  or should i  remove the `start_leader_counter` metric check directly?

**Comment by lance6716** [comment]
> need mater1 to became the leader

maybe you need `operate-leader evict`

**Comment by lichunzhu** [comment]
Could you please also help fix https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/241/pipeline/1800?

**Comment by Ehco1996** [comment]
> Could you please also help fix https://ci.pingcap.net/blue/organizations/jenkins/dm_ghpr_new_test/detail/dm_ghpr_new_test/241/pipeline/1800?

@lichunzhu  got it

**Comment by Ehco1996** [comment]
> > need mater1 to became the leader
> 
> maybe you need `operate-leader evict`

@lance6716 i think still need sleep here, `operate-leader evict` don't work beacuse master2 may campagin leader before `operate-leader evict`

File: tests/ha/run.sh [hunk 1]
```diff
@@ -44,10 +46,6 @@ function run() {
 	check_rpc_alive $cur/../bin/check_master_online 127.0.0.1:$MASTER_PORT3
 	check_metric $MASTER_PORT3 'start_leader_counter' 3 -1 1 # master3 is not leader
 
-	# check worker state
-	check_metric $MASTER_PORT1 'dm_master_worker_state{worker="worker1"}' 3 1 3
-	check_metric $MASTER_PORT1 'dm_master_worker_state{worker="worker2"}' 3 1 3
```
Comment (Ehco1996) [inline tests/ha/run.sh:49]: delete this check beause don''t know who is leader but `dm_master_worker_state` metric only in leader's metrics

**Comment by Ehco1996** [comment]
@lichunzhu @lance6716  replay three times in jenkins all tests passed, please cc

File: tests/ha/run.sh [hunk 2]
```diff
@@ -19,8 +19,10 @@ function run() {
 	run_dm_master $WORK_DIR/master2 $MASTER_PORT2 $cur/conf/dm-master2.toml
 	check_rpc_alive $cur/../bin/check_master_online 127.0.0.1:$MASTER_PORT1
 	check_rpc_alive $cur/../bin/check_master_online 127.0.0.1:$MASTER_PORT2
-	check_metric $MASTER_PORT1 'start_leader_counter' 3 0 2
-	check_metric $MASTER_PORT2 'start_leader_counter' 3 -1 1 # master2 is not leader
+
+	# master1 or master2 join campaign
```
Comment (lance6716) [inline tests/ha/run.sh:23]: could it be 
check_metric $MASTER_PORT1 'start_leader_counter' 3 0 2 || check_metric $MASTER_PORT1 'start_leader_counter' 3 0 2

**Review by lance6716 — COMMENTED** [review]
rest lgtm

File: tests/ha/run.sh (hunk 2 above)
Comment (lance6716) [reply tests/ha/run.sh:23]: `3 -1 2` is too weak IMO

File: tests/ha/run.sh (hunk 1 above)
Comment (lance6716) [reply tests/ha/run.sh:49]: check_metric $MASTER_PORT1 'dm_master_worker_state{worker="worker1"}' 3 1 3 || check_metric $MASTER_PORT2 'dm_master_worker_state{worker="worker1"}' 3 1 3

File: tests/ha/run.sh (hunk 2 above)
Comment (Ehco1996) [reply tests/ha/run.sh:23]: but the metrics can be  0 or 1 ...

File: tests/_utils/wait_process_exit [hunk 3]
```diff
@@ -1,15 +1,35 @@
...
+# Counter to keep count of how many seconds have passed
+count=0
+
+PID=$(pgrep $process)
+if [ -z "$PID" ]; then
+	echo "process $process already exit"
+	exit
+fi
+
+while kill $PID >/dev/null; do
```
Comment (glorv) [inline tests/_utils/wait_process_exit:17]: Do we actually need to send kill signal multi times? I think kill once and wait in a loop is enough

**Review by glorv — COMMENTED** [review]
rest LGTM

File: tests/_utils/wait_process_exit (hunk 3 above)
Comment (Ehco1996) [reply tests/_utils/wait_process_exit]: fixed

**Review by GMHDBJD — APPROVED** [review]
/lgtm

**Comment by lance6716** [comment]
please fix test

```
[ERROR] [main.go:89] ["fail to start dm-worker"] [error="[code=40048:class=dm-worker:scope=internal:level=high], Message: start server, RawCause: listen tcp 0.0.0.0:8262: bind: address already in use"]
```

File: tests/_utils/wait_process_exit [hunk 4]
```diff
@@ -2,37 +2,21 @@
...
+while true; do
+	pgrep $process >/dev/null 2>&1
+	ret=$?
+	if [ "$ret" != "0" ]; then
+		echo "process $process already exit"
+		break
+	fi
+	sleep 0.2
+	echo "wait process $process exit..."
+done
```
Comment (lance6716) [inline tests/_utils/wait_process_exit:14]: seems we don't need lines after line 15

Comment (Ehco1996) [reply tests/_utils/wait_process_exit:14]: i not sure but i found in some case , process still running after `wait_process_exit`,so i add `kill -9` to guarantee it

Comment (lance6716) [reply tests/_utils/wait_process_exit:14]: when `while true` break, line 17 is always true 🤔

Comment (Ehco1996) [reply tests/_utils/wait_process_exit:27]: 🙄  you are right, move `kill -9` inside loop

Comment (lance6716) [reply tests/_utils/wait_process_exit:27]: after kill -9, still need to wait processs exit

**Review by lichunzhu — COMMENTED** [review]
Rest LGTM

File: tests/_utils/wait_process_exit [hunk 5]
```diff
@@ -11,5 +20,10 @@ while true; do
 		break
 	fi
 	sleep 0.2
+	((count++))
 	echo "wait process $process exit..."
+	if [ $count -gt $WAIT_COUNT ] && [ $KILL_TIMES -lt 2 ]; then
+		((KILL_ONCE++))
```
[the remark this answers is not in the document]
Comment (Ehco1996) [reply tests/_utils/wait_process_exit]: it's a typo... should be `KILL_TIMES`,fixed

File: tests/_utils/wait_process_exit [hunk 6]
```diff
@@ -1,8 +1,17 @@
...
 process=$1
 
+PID=$(pgrep $process)
+if [ -z "$PID" ]; then
+	echo "process $process already exit"
+	exit
+fi
+
+WAIT_COUNT=10
+KILL_TIMES=1 # make sure only send kill -9 once.
```
Comment (lance6716) [inline tests/_utils/wait_process_exit:12]: why it's initialized as 1 🤔

**Review by lance6716 — APPROVED** [review]

File: tests/_utils/wait_process_exit (hunk 6 above)
Comment (Ehco1996) [reply tests/_utils/wait_process_exit:12]: 🤡  🤡

**Review by GMHDBJD — APPROVED** [review]
/lgtm

**Pull request merged**